adam_core/coordinates/variants.py

```diff
@@ -0,0 +1,215 @@
+from typing import Literal, Tuple, Union
+
+import numpy as np
+import quivr as qv
+
+from .cartesian import CartesianCoordinates
+from .cometary import CometaryCoordinates
+from .covariances import (
+    mean_and_covariance_from_weighted_samples,
+    sample_covariance_random,
+    sample_covariance_sigma_points,
+)
+from .keplerian import KeplerianCoordinates
+from .spherical import SphericalCoordinates
+
+CoordinateType = Union[
+    CartesianCoordinates,
+    KeplerianCoordinates,
+    CometaryCoordinates,
+    SphericalCoordinates,
+]
```

Comment:
Is it really sensible to do this on Cometary and Keplerian elements? That seems like it could be super non-gaussian. Like what does it mean to symmetrically sample from the eccentricity? Just seems risky. I'm a fan of only working with cartesian coordinates internally.

Reply:
This was my only question as well. Does one get less accurate results by forcing the user to convert to Cartesian first (less true to the original coordinates)?

Reply:
> Is it really sensible to do this on Cometary and Keplerian elements? That seems like it could be super non-gaussian. Like what does it mean to symmetrically sample from the eccentricity? Just seems risky. I'm a fan of only working with cartesian coordinates internally.

This is the beauty of sampling. What we really care about is mapping volumes from one space to another and the way we can do this is using sampling methods -- the whole point of the addition of variants! Here are some plots where I've defined a Keplerian orbit with a covariance matrix:
```python
keplerian_coordinates = KeplerianCoordinates.from_kwargs(
    time=Times.from_astropy(
        Time([59000], scale="tdb", format="mjd")
    ),
    a=[2.0],
    e=[0.8],
    i=[10.0],
    raan=[60.0],
    ap=[40.0],
    M=[30.0],
    covariance=CoordinateCovariances.from_sigmas(np.array([[0.1, 0.01, 1, 0.1, 0.01, 0.01]])),
    origin=Origin.from_kwargs(code=["SUN"]),
    frame="ecliptic"
)
```
We can then do two things. 
1. Take the input Keplerian orbit and covariance, sample the Keplerian covariance and then convert each sample to Cartesian and plot those (grey points). 
2. Take the Keplerian orbit and convert it to Cartesian and then sample the Cartesian covariance for variants (red points). 

What we really care about is to make sure that the underlying shapes plotted are approximately the same. Because it means the underlying covariance transformation preserved the shape of the original distribution. This is really what we care about and want to be able to study. There will of course be edgecases like our code will eventually need to be able to handle creating variants at e=0.9 +- 0.2. It needs to know that when e > 1 then a < 0. It doesn't do that now but that is something we can add later when we need it. The conversion code will assert an error when orbits are incorrectly defined. 

![keplerian_vs_cartesian_monte-carlo](https://github.com/B612-Asteroid-Institute/adam_core/assets/4206654/4a41e0c7-d8f8-4605-955e-d60702dde0b9)

Also to showcase the insane power of sigma-points sampling. Look how beautiful this is: 
![keplerian_vs_cartesian_sigma-point](https://github.com/B612-Asteroid-Institute/adam_core/assets/4206654/bd139ebf-5b1a-4d07-a2b3-7cc6e47c6c8a)

Reply:
> What we really care about is to make sure that the underlying shapes plotted are approximately the same. Because it means the underlying covariance transformation preserved the shape of the original distribution. This is really what we care about and want to be able to study. 

Yes! These plots are very encouraging, for sure.

> There will of course be edgecases like our code will eventually need to be able to handle creating variants at e=0.9 +- 0.2. It needs to know that when e > 1 then a < 0. It doesn't do that now but that is something we can add later when we need it. The conversion code will assert an error when orbits are incorrectly defined.

Right, this is the sort of issue that I was worried about. You've described a crucial one, which is that the covariance does not really encode complete information about Keplerian parameterizations. Since sigma points rely upon marginal distributions, I'm concerned that we're going to be in for some trouble; forbidden regions for the full vector is a great example.

Another genre of concern though is that this implementation of sigma points assumes that symmetric samples are equally valid: that `e+σ` and `e-σ` should be equally weighted, and are equally representative of the marginal distribution of e. That is, they assume higher-order moments are all zero. I'm concerned that that could be _very_ untrue for us. I'd expect that the likelihood function for `e` has very non-negligible skewness.

(Actually, we should probably be equally concerned for Cartesian parameterizations. The only real symmetric gaussians lying around are probably the astrometric uncertainties of sky-plane coordinates.)

The issue with higher-order moments isn't necessarily fatal. [Julier 1998](https://ui.adsabs.harvard.edu/abs/1998SPIE.3373..271J/abstract)[^1] (cited by van der Merwe & Wan on this exact subject) looks promising, and suggests that you simply need more sigma points. And [Julier & Uhlmann 2004](https://www.cs.ubc.ca/~murphyk/Papers/Julier_Uhlmann_mar04.pdf) has a lot to say about skewness and kurtosis in section 5, "Mitigating Unmodeled Effects".

This is all me running on fear/uncertainty/doubt, though, and if we set `alpha` to a small enough value, maybe we'll be just fine. The plots are definitely very encouraging!

[^1]: Direct link: http://www.gatsby.ucl.ac.uk/~byron/paper.pdf

Reply:
So I'm playing around with different alpha values. The issue with alpha < 1 is that we get a negative weight for the mean state which is unsupported in `np.cov` so I'll update the code to not use that and instead do the linalg manually.

The second thing is that the unit tests of covariance reconstruction actually aren't testing the real-life use case. We haven't applied a non-linear transformation and are just testing the reconstruction of the input covariance which is probably biasing us to the wrong set of parameters. Setting alpha = 1e-2, beta = 2, kappa = 0 doesn't reproduce the input covariance as well as alpha = 1, beta = 0, kappa = 0, but it still does a decent job (differences at the 1e-15 vs 1e-25 level, basically 0 for both). You are correct that we will likely want the former as it captures more moments of the distribution (since beta is a function of the kurtosis in the case of Gaussians). 

I'll add a pathway for the additional mitigations from [Julier & Uhlmann 2004](https://www.cs.ubc.ca/~murphyk/Papers/Julier_Uhlmann_mar04.pdf) since they are fairly straightforward for kurtosis. I need to let the statements about skewness marinate a little more -- it's not obvious we can do much there or at least find a one-size-fits-all solution. 

I'm going to document this a bit for the THOR use case since we will want to investigate this further which we can do with simulation. What we have could be good enough as it is for observation gathering but it's better to be sure.

Reply:
I'm going to continue this part of the work in a separate PR as this one is large enough.